src/main/core/work/task.rs

```diff
@@ -0,0 +1,173 @@
+use std::sync::Arc;
+
+use atomic_refcell::AtomicRefCell;
+
+use crate::{
+    core::worker::Worker,
+    host::host::Host,
+    utility::{IsSend, IsSync},
+};
+
+/// Mostly for interoperability with C APIs.
+/// In Rust code that doesn't need to interact with C, it may make more sense
+/// to directly use a `FnMut(&mut Host)` trait object.
+pub struct TaskRef {
+    inner: Arc<AtomicRefCell<dyn FnMut(&mut Host) + Send + Sync>>,
+    // Runtime memory error checking to help catch errors that C code is prone
+    // to.  Can probably drop once C interop is removed, if we don't drop this
+    // type altogether.
+    #[cfg(debug_assertions)]
+    magic: u32,
+}
+
+impl TaskRef {
+    #[cfg(debug_assertions)]
+    const MAGIC: u32 = 0xe0408897;
+
+    pub fn new<T: 'static + FnMut(&mut Host) + Send + Sync>(f: T) -> Self {
+        Worker::increment_object_alloc_counter("TaskRef");
+        Self {
+            inner: Arc::new(AtomicRefCell::new(f)),
+            #[cfg(debug_assertions)]
+            magic: Self::MAGIC,
+        }
+    }
+
+    pub fn execute(&mut self, host: &mut Host) {
+        self.check_magic();
+        let mut inner = self.inner.borrow_mut();
+        inner(host)
+    }
+
+    #[cfg(debug_assertions)]
+    fn check_magic(&self) {
+        debug_assert!(self.magic == Self::MAGIC);
+    }
+    #[cfg(not(debug_assertions))]
+    fn check_magic(&mut self) {}
+
+    #[cfg(debug_assertions)]
+    fn clear_magic(&mut self) {
+        unsafe { std::ptr::write_volatile(&mut self.magic, 0) };
+    }
+    #[cfg(not(debug_assertions))]
+    fn clear_magic(&mut self) {}
+}
+
+impl Drop for TaskRef {
+    fn drop(&mut self) {
+        self.check_magic();
+        self.clear_magic();
+        Worker::increment_object_dealloc_counter("TaskRef");
+    }
+}
+
+impl Clone for TaskRef {
+    fn clone(&self) -> Self {
+        Worker::increment_object_alloc_counter("TaskRef");
+        Self {
+            inner: self.inner.clone(),
+            #[cfg(debug_assertions)]
+            magic: self.magic.clone(),
+        }
+    }
+}
+
+impl IsSend for TaskRef {}
+impl IsSync for TaskRef {}
+
+pub mod export {
+    use super::*;
+
+    use crate::{cshadow, host::host::Host, utility::SyncSendPointer};
+
+    pub type TaskCallbackFunc =
+        extern "C" fn(*mut cshadow::Host, *mut libc::c_void, *mut libc::c_void);
+    pub type TaskObjectFreeFunc = Option<extern "C" fn(*mut libc::c_void)>;
+    pub type TaskArgumentFreeFunc = Option<extern "C" fn(*mut libc::c_void)>;
+
+    struct CTask {
+        callback: TaskCallbackFunc,
+        object: SyncSendPointer<libc::c_void>,
+        argument: SyncSendPointer<libc::c_void>,
+        object_free: TaskObjectFreeFunc,
+        argument_free: TaskArgumentFreeFunc,
+    }
+
+    impl CTask {
+        /// SAFETY: Objects pointed to by `object` and `argument` must be
+        /// `Send`.
+        unsafe fn new(
+            callback: TaskCallbackFunc,
```

Comment:
I'm trying to think about the more global safety requirements here. So in Shadow we assume a host object and all of its "sub-objects" are only ever accessed by one thread at a time. When executing a managed process, we hold the host's lock so that we have exclusive control of the host while it's running and while we're executing syscall handlers. But this host lock doesn't protect shared/ref-counted objects. Ideally only objects that are owned by the host should hold references to a shared object belonging to that same host. For example, the host's network interface can hold a shared reference to a socket object belonging to the same host since to access the host's network interface, you must already be holding the host's lock.

(The shared/ref-counted objects have an `AtomicRefCell` to panic in the case of simultaneous mutable borrows, but that should be seen as a last-resort check, and we should aim to never put ourselves in a situation where that could happen.)

A task is an exception here since in Shadow it can contain a reference to a host's shared object (for example a shared socket reference), but the task is owned by the scheduler which exists outside of the host and is not protected by the host's lock. Since a task is shared between two threads (the scheduler thread and the thread that a host is running on), we should either make sure that the task never contains a shared/ref-counted object that belongs to a host, or we should make sure a given task only contains shared objects belonging to a single host. Since the task callback takes a `&Host` as an argument, it must mean that we have acquired the lock for that host, but we should also make sure that the task's closure only contains objects that belong to that host and no other hosts.

I don't think there is any way to enforce this. Maybe in the comments we should specify that the closure must compare the passed host argument with the owning host of each shared object contained in the closure to make sure they're equivalent. But currently no existing shared objects (such as sockets/pipes/etc) know anything about the host they belong to. So maybe these objects should also store the host ID, and then every closure can compare the host ID of the shared object to the host ID of the provided host?

I'm not sure if there's a good way to handle this, other than documenting something here.

Reply:
Fleshed out the comments a bit more, especially for `taskref_new`, and added a runtime check that the callbacks expected in the context of the expected Host